base/collection_xattr.go

```diff
@@ -428,68 +452,105 @@ func (c *Collection) SubdocSetXattrs(k string, xvs map[string][]byte) (casOut ui
 	if !c.IsSupported(sgbucket.BucketStoreFeatureMultiXattrSubdocOperations) && len(xattrs) >= 2 {
 		return 0, fmt.Errorf("UpdateXattrs: more than 1 xattr %v not supported in UpdateXattrs in this version of Couchbase Server", xattrs)
 	}
 	c.Bucket.waitForAvailKvOp()
 	defer c.Bucket.releaseKvOp()
 
-	mutateOps := make([]gocb.MutateInSpec, 0, len(xattrs))
-	for xattrKey, xattrVal := range xattrs {
-		mutateOps = append(mutateOps, gocb.UpsertSpec(xattrKey, bytesToRawMessage(xattrVal), UpsertSpecXattr))
+	mutateOps, err := getUpsertSpecs(xattrs)
+	if err != nil {
+		return 0, err
+	}
+	for _, xattrKey := range xattrsToDelete {
+		if _, ok := xattrs[xattrKey]; ok {
+			return 0, fmt.Errorf("%s: %w", xattrKey, sgbucket.ErrUpsertAndDeleteSameXattr)
+		}
+		mutateOps = append(mutateOps, gocb.RemoveSpec(xattrKey, RemoveSpecXattr))
 	}
 	mutateOps = appendMacroExpansions(mutateOps, opts)
 
 	options := &gocb.MutateInOptions{
 		Expiry:        CbsExpiryToDuration(exp),
-		StoreSemantic: gocb.StoreSemanticsUpsert,
+		StoreSemantic: gocb.StoreSemanticsReplace,
 		Cas:           gocb.Cas(cas),
 	}
+
 	options.Internal.DocFlags = gocb.SubdocDocFlagAccessDeleted
 	fillMutateInOptions(ctx, options, opts)
 
 	result, mutateErr := c.Collection.MutateIn(k, mutateOps, options)
 	if mutateErr != nil {
+		if isKVError(mutateErr, memd.StatusSubDocBadMulti) {
+			mutateErr = fmt.Errorf("%w: %w", ErrXattrNotFound, mutateErr)
+		}
 		return 0, mutateErr
 	}
 	return uint64(result.Cas()), nil
 }
 
 // updateBodyAndXattr updates the document body and xattrs of an existing document. Writes cas and crc32c to the xattr using macro expansion.
-func (c *Collection) updateBodyAndXattrs(ctx context.Context, k string, exp uint32, cas uint64, opts *sgbucket.MutateInOptions, v interface{}, xattrs map[string][]byte) (casOut uint64, err error) {
+func (c *Collection) updateBodyAndXattrs(ctx context.Context, k string, exp uint32, cas uint64, opts *sgbucket.MutateInOptions, v interface{}, xattrs map[string][]byte, xattrsToDelete []string) (casOut uint64, err error) {
 	c.Bucket.waitForAvailKvOp()
 	defer c.Bucket.releaseKvOp()
 
-	mutateOps := make([]gocb.MutateInSpec, 0, len(xattrs)+1)
-	for xattrKey, xattrVal := range xattrs {
-		mutateOps = append(mutateOps, gocb.UpsertSpec(xattrKey, bytesToRawMessage(xattrVal), UpsertSpecXattr))
+	mutateOps, err := getUpsertSpecs(xattrs)
+	if err != nil {
+		return 0, err
+	}
+	for _, xattrKey := range xattrsToDelete {
+		if _, ok := xattrs[xattrKey]; ok {
+			return 0, fmt.Errorf("%s: %w", xattrKey, sgbucket.ErrUpsertAndDeleteSameXattr)
+		}
+		mutateOps = append(mutateOps, gocb.RemoveSpec(xattrKey, RemoveSpecXattr))
+
 	}
 	mutateOps = append(mutateOps, gocb.ReplaceSpec("", bytesToRawMessage(v), nil))
 	mutateOps = appendMacroExpansions(mutateOps, opts)
 
 	options := &gocb.MutateInOptions{
 		Expiry:        CbsExpiryToDuration(exp),
-		StoreSemantic: gocb.StoreSemanticsUpsert,
+		StoreSemantic: gocb.StoreSemanticsReplace,
```

Comment:
Same question as above - do the xattrs have to already exist to use StoreSemanticsReplace?